Include/internal/pycore_stackref.h

```diff
@@ -8,187 +8,269 @@ extern "C" {
 #  error "this header requires Py_BUILD_CORE define"
 #endif
 
+#include "pycore_object_deferred.h"
+
 #include <stddef.h>
 
+/*
+  Authors: Ken Jin, Mark Shannon
+
+  This file introduces a new API for handling references on the stack, called
+  _PyStackRef. This API is inspired by HPy.
+
+  There are 3 main operations, that convert _PyStackRef to PyObject* and
+  vice versa:
+
+    1. Borrow (discouraged)
+    2. Steal
+    3. New
+
+  Borrow means that the reference is converted without any change in ownership.
+  This is discouraged because it makes unboxed integers harder in the future.
+
+  Steal "steals" the reference. What this means is that the old reference
+  is now invalid, and the reference is now owned by what it is converted to.
+
+  New creates a new reference from the old reference. The old reference
+  is unchanged.
+
+  With these 3 API, a strict stack discipline must be maintained. All
+  _PyStackRef must be operated on by the new reference operations:
+
+    1. DUP
+    2. CLOSE
+    3. CLEAR
+
+   DUP is equivalent to Py_NewRef. It creates a new reference from an old
```

Comment:
It isn't quite equivalent. It is similar, though.
The difference is that we are now dealing with references, whereas before we were dealing with reference counts.

We don't enforce this, but this code is incorrect: (HPy can enforce this in debug mode).
```
    PyStackRef_DUP(x);
    y = x;
```
but this was legal, if poor style:
```
     Py_NewRef(x);
    y = x;
```